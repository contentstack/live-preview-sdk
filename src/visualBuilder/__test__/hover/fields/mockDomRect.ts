export const mockDomRect = {
    singleHorizontal: () => ({
        bottom: 88.3984375,
        height: 54.3984375,
        left: 34,
        right: 862,
        top: 34,
        width: 828,
        x: 34,
        y: 34,
    }),
    singleLeft: () => ({
        x: 51,
        y: 51,
        width: 27.7734375,
        height: 20.3984375,
        top: 51,
        bottom: 71.3984375,
        left: 51,
        right: 78.7734375,
    }),
    singleRight: () => ({
        x: 110.7734375,
        y: 51,
        width: 53.59375,
        height: 20.3984375,
        top: 51,
        bottom: 71.3984375,
        left: 110.7734375,
        right: 164.3671875,
    }),
    singleVertical: () => ({
        bottom: 195.1953125,
        height: 90.796875,
        left: 34,
        right: 862,
        top: 104.3984375,
        width: 828,
        x: 34,
        y: 104.3984375,
    }),
    singleTop: () => ({
        x: 51,
        y: 121.3984375,
        width: 794,
        height: 20.3984375,
        top: 121.3984375,
        bottom: 141.796875,
        left: 51,
        right: 845,
    }),
    singleBottom: () => ({
        x: 51,
        y: 157.796875,
        width: 794,
        height: 20.3984375,
        top: 157.796875,
        bottom: 178.1953125,
        left: 51,
        right: 845,
    }),
    multipleLeft: () => ({
        x: 51,
        y: 228.1953125,
        width: 87.59375,
        height: 90.796875,
        top: 228.1953125,
        bottom: 318.9921875,
        left: 51,
        right: 138.59375,
    }),
    multipleRight: () => ({
        x: 170.59375,
        y: 228.1953125,
        width: 87.59375,
        height: 90.796875,
        top: 228.1953125,
        bottom: 318.9921875,
        left: 170.59375,
        right: 258.1875,
    }),
    multipleTop: () => ({
        x: 51,
        y: 368.9921875,
        width: 794,
        height: 90.796875,
        top: 368.9921875,
        bottom: 459.7890625,
        left: 51,
        right: 845,
    }),
    multipleBottom: () => ({
        x: 51,
        y: 475.7890625,
        width: 794,
        height: 90.796875,
        top: 475.7890625,
        bottom: 566.5859375,
        left: 51,
        right: 845,
    }),
    multipleChildTop: () => ({
        x: 68,
        y: 385.9921875,
        width: 760,
        height: 20.3984375,
        top: 385.9921875,
        bottom: 406.390625,
        left: 68,
        right: 828,
    }),
    multipleChildBottom: () => ({
        x: 68,
        y: 422.390625,
        width: 760,
        height: 20.3984375,
        top: 422.390625,
        bottom: 442.7890625,
        left: 68,
        right: 828,
    }),
};
