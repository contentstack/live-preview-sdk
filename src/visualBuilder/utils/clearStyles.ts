const originalStyles = new WeakMap<
    HTMLElement,
    {
        visibility?: string;
        transition?: string;
        animation?: string;
    }
>();

export function clearVisibelityStyles(element: HTMLElement) {
    const originalStyleValues = {
        visibility: element.style.visibility,
        transition: element.style.transition,
        animation: element.style.animation,
    };

    originalStyles.set(element, originalStyleValues);

    element.style.visibility = "hidden";
    element.style.transition = "none";
    element.style.animation = "none";
}

export function restoreVisibelityStyles(element: HTMLElement) {
    const storedStyles = originalStyles.get(element);

    if (storedStyles) {
        element.style.visibility = storedStyles.visibility || "";
        element.style.transition = storedStyles.transition || "";
        element.style.animation = storedStyles.animation || "";

        originalStyles.delete(element);
    }
}
