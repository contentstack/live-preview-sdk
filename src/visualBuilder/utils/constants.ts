import { FieldDataType } from "./types/index.types";

export const numericInputRegex = /^-?\d*(\.\d*)?([eE][-+]?\d*)?$/;

export const VISUAL_BUILDER_FIELD_TYPE_ATTRIBUTE_KEY = "data-cslp-field-type";

export const VISUAL_BUILDER_CHANNEL_ID = "visual-builder";

export const LIVE_PREVIEW_OUTLINE_WIDTH_IN_PX = 2;

// Minimum distance from top edge to prevent toolbar from being hidden
export const TOP_EDGE_BUFFER = 42;
export const RIGHT_EDGE_BUFFER = 180;

export const TOOLBAR_EDGE_BUFFER = 8;

export const DATA_CSLP_ATTR_SELECTOR = "data-cslp";

/**
 * The field that can be directly modified using contenteditable=true.
 * This includes all text fields like title and numbers.
 */
export const ALLOWED_INLINE_EDITABLE_FIELD: FieldDataType[] = [
    FieldDataType.SINGLELINE,
    FieldDataType.MULTILINE,
    FieldDataType.NUMBER,
];

export const ALLOWED_MODAL_EDITABLE_FIELD: FieldDataType[] = [
    FieldDataType.HTML_RTE,
    FieldDataType.MARKDOWN_RTE,
    FieldDataType.JSON_RTE,
    FieldDataType.CUSTOM_FIELD,
    FieldDataType.LINK,
    FieldDataType.ISODATE,
    FieldDataType.URL,
];

export const ALLOWED_REPLACE_FIELDS: FieldDataType[] = [
    FieldDataType.REFERENCE,
    FieldDataType.FILE,
];

export const DEFAULT_MULTIPLE_FIELDS: FieldDataType[] = [
    FieldDataType.GLOBAL_FIELD,
    FieldDataType.GROUP,
    FieldDataType.BLOCK,
];

export const mentionLimit = 20;
export const maxMessageLength = 500;
