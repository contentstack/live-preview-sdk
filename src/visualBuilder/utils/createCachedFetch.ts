/**
 * Creates a cached async fetch function with support for any number of arguments
 * @param fetchFn - The async function to cache
 * @param uidResolver - Function that generates a unique ID from the arguments passed to fetchFn
 * @returns A cached version of the fetch function with the same signature
 */
export function createCachedFetch<TArgs extends any[], TResult>(
    fetchFn: (...args: TArgs) => Promise<TResult>,
    uidResolver: (...args: TArgs) => string = (...args) => JSON.stringify(args)
): {
    (...args: TArgs): Promise<TResult>;
    clearCache: () => void;
} {
    // Cache storage
    const cache = new Map<string, TResult>();
    // Track in-flight requests
    const pendingPromises = new Map<string, Promise<TResult>>();

    /**
     * The cached fetch function
     * @param args - Arguments to pass to the original fetch function
     * @returns Promise that resolves with the data
     */
    async function cachedFetch(...args: TArgs): Promise<TResult> {
        // Generate unique ID for these arguments
        const uid = uidResolver(...args);

        // Return cached value if available
        if (cache.has(uid)) {
            return cache.get(uid)!;
        }

        // Return existing promise if request is already in progress
        if (pendingPromises.has(uid)) {
            return pendingPromises.get(uid)!;
        }

        // Create new promise for this request
        const promise = fetchFn(...args)
            .then((data) => {
                // Store result in cache
                cache.set(uid, data);
                // Remove from pending
                pendingPromises.delete(uid);
                return data;
            })
            .catch((error) => {
                // Clean up on error
                pendingPromises.delete(uid);
                throw error;
            });

        // Store the promise
        pendingPromises.set(uid, promise);
        return promise;
    }

    // Add clearCache method to the function
    cachedFetch.clearCache = () => {
        cache.clear();
        pendingPromises.clear();
    };

    return cachedFetch;
}
