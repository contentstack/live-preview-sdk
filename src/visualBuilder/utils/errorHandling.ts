
export function hasPostMessageError(obj: any): boolean {
    return obj?.error === true;
}