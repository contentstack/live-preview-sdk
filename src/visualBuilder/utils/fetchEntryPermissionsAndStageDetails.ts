import { getEntryPermissionsCached } from "./getEntryPermissionsCached";
import { getWorkflowStageDetails } from "./getWorkflowStageDetails";

export async function fetchEntryPermissionsAndStageDetails({
    entryUid,
    contentTypeUid,
    locale,
    variantUid,
}: {
    entryUid: string;
    contentTypeUid: string;
    locale: string;
    variantUid?: string | undefined;
}) {
    const entryAclPromise = getEntryPermissionsCached({
        entryUid,
        contentTypeUid,
        locale,
    });
    const entryWorkflowStageDetailsPromise = getWorkflowStageDetails({
        entryUid,
        contentTypeUid,
        locale,
        variantUid,
    });
    const results = await Promise.allSettled([
        entryAclPromise,
        entryWorkflowStageDetailsPromise,
    ]);
    if (results[0].status === "rejected") {
        console.debug(
            "[Visual Builder] Error retrieving entry permissions",
            results[0].reason
        );
    }
    if (results[1].status === "rejected") {
        console.debug(
            "[Visual Builder] Error retrieving entry stage details",
            results[1].reason
        );
    }
    const acl =
        results[0].status === "fulfilled" ? results[0].value : undefined;
    const workflowStage =
        results[1].status === "fulfilled" ? results[1].value : undefined;
    return {
        acl,
        workflowStage,
    };
}
