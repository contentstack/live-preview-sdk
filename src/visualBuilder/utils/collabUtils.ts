import {
    ICommentState,
    IMentionedList,
    IMentionItem,
    IMessageDTO,
    IUserDTO,
    IUserState,
} from "../types/collab.types";
import { maxMessageLength, mentionLimit } from "./constants";
import { uniqBy } from "lodash-es";
import DOMPurify from "dompurify";

const escapeRegExp = (string: string): string => {
    return string.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
};

/**
 * Generates the title for the thread based on the number of comments.
 * @param {number} commentCount - The number of comments.
 * @returns {string} The title for the thread.
 */
export const getThreadTitle = (commentCount: number): string => {
    if (commentCount === 0) return "Add New Feedback";
    return commentCount === 1 ? "1 Feedback" : `${commentCount} Feedbacks`;
};

/**
 * returns the available email.
 * @param {IUserDTO} user - The user object.
 * @returns {string} The user's email.
 */
export const getUserName = (user: IUserDTO): string => {
    return user.email;
};

/**
 * Validates the comment length and the number of mentions.
 * @param {string} comment - The comment message.
 * @param {IMentionedList} toUsers - The list of mentioned users.
 * @returns {string} The error message if validation fails, otherwise an empty string.
 */
export const validateCommentAndMentions = (
    comment: string,
    toUsers: IMentionedList
): string => {
    if (comment.length > maxMessageLength) {
        return `Limit exceeded. You can have a maximum length of ${maxMessageLength} characters.`;
    }
    if (toUsers.length > mentionLimit) {
        return `Limit exceeded. You can tag a maximum of ${mentionLimit} users.`;
    }
    return "";
};

/**
 * Removes mentions that no longer exist in the message.
 * @param {string} message - The comment message.
 * @param {IMentionedList} toUsers - The list of mentioned users.
 * @returns {Object} The updated lists of mentioned users.
 */
export const filterOutInvalidMentions = (
    message: string,
    toUsers: IMentionedList
) => {
    const to_users_temp = toUsers.filter((user) =>
        message.includes(user.display)
    );

    return {
        toUsers: uniqBy(to_users_temp, "id"),
    };
};

/**
 * Replaces mention placeholders with display names in the comment message.
 * @param {IMessageDTO | undefined} comment - The comment object.
 * @param {IUserState} userState - The user state containing user and role maps.
 * @param {"text" | "html"} profile - The format for the output message, either plain text or HTML.
 * @returns {string | undefined} The formatted message or undefined if the comment is not provided.
 */
export const getMessageWithDisplayName = (
    comment: IMessageDTO | undefined | null,
    userState: IUserState,
    profile: "text" | "html"
): string | undefined => {
    if (!comment) return undefined;

    let tempText = comment.message;

    comment?.toUsers?.forEach((user) => {
        const userPattern = new RegExp(`{{${user}}}`, "g");
        const userData = userState.userMap[user];
        const replacement =
            profile === "html"
                ? `<b>@${userData.display || getUserName(userData)}</b>`
                : `@${userData.display || getUserName(userData)}`;
        tempText = tempText.replace(userPattern, replacement);
    });

    return tempText;
};

/**
 * Sanitizes HTML content to prevent XSS attacks.
 * @param {any} dirty - The unsanitized HTML content.
 * @returns {string} The sanitized HTML content.
 */
export const sanitizeData = (dirty: any): string => {
    return DOMPurify.sanitize(dirty, { USE_PROFILES: { html: true } });
};

/**
 * Constructs the comment body with mentions replaced by their unique identifiers.
 * @param {ICommentState} state - The state containing the comment and mentions.
 * @returns {Object} The comment body containing the sanitized message and mentioned users.
 */
export const getCommentBody = (state: ICommentState): ICommentState => {
    let finalMessage = state.message.trim().replace(/\s+/g, " ");
    const comment = {
        message: finalMessage,
        toUsers: [],
        images: [],
        createdBy: state.createdBy,
        author: state.author,
    };

    const updateMentionToUID = (
        entity: IMentionItem,
        result: Array<string>
    ) => {
        const displayName = entity.display;

        const escapedDisplayName = escapeRegExp(displayName);
        const regexUser = new RegExp(escapedDisplayName, "g");
        finalMessage = finalMessage.replace(regexUser, `{{${entity.id}}}`);
        result.push(entity.display);
    };

    state.toUsers?.forEach((user) => updateMentionToUID(user, comment.toUsers));

    comment.message = finalMessage;
    return comment;
};
