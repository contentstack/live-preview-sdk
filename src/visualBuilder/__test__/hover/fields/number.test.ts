import { screen } from "@testing-library/preact";
import { getFieldSchemaMap } from "../../../../__test__/data/fieldSchemaMap";
import { sleep, waitForHoverOutline } from "../../../../__test__/utils";
import Config from "../../../../configManager/configManager";
import { VisualBuilder } from "../../../index";
import { FieldSchemaMap } from "../../../utils/fieldSchemaMap";
import { mockDomRect } from "./mockDomRect";

vi.mock("../../../utils/visualBuilderPostMessage", async () => {
    const { getAllContentTypes } = await vi.importActual<
        typeof import("../../../../__test__/data/contentType")
    >("../../../../__test__/data/contentType");
    const contentTypes = getAllContentTypes();
    return {
        __esModule: true,
        default: {
            send: vi.fn().mockImplementation((eventName: string) => {
                if (eventName === "init")
                    return Promise.resolve({
                        contentTypes,
                    });
                return Promise.resolve();
            }),
        },
    };
});

global.ResizeObserver = vi.fn().mockImplementation(() => ({
    observe: vi.fn(),
    unobserve: vi.fn(),
    disconnect: vi.fn(),
}));

describe("When an element is hovered in visual builder mode", () => {
    let mousemoveEvent: Event;

    beforeAll(() => {
        FieldSchemaMap.setFieldSchema(
            "all_fields",
            getFieldSchemaMap().all_fields
        );
    });

    beforeEach(() => {
        Config.reset();
        Config.set("mode", 2);
        mousemoveEvent = new Event("mousemove", {
            bubbles: true,
            cancelable: true,
        });
    });

    afterEach(() => {
        vi.clearAllMocks();
        document.getElementsByTagName("html")[0].innerHTML = "";
    });

    afterAll(() => {
        Config.reset();
    });

    describe("number field", () => {
        let numberField: HTMLParagraphElement;
        let visualBuilder: VisualBuilder;

        beforeEach(() => {
            numberField = document.createElement("p");
            numberField.setAttribute(
                "data-cslp",
                "all_fields.bltapikey.en-us.number"
            );

            numberField.getBoundingClientRect = vi
                .fn()
                .mockReturnValue(mockDomRect.singleLeft());

            document.body.appendChild(numberField);

            visualBuilder = new VisualBuilder();
        });

        afterEach(() => {
            visualBuilder.destroy();
        });

        test("should have outline and custom cursor", async () => {
            numberField.dispatchEvent(mousemoveEvent);
            await waitForHoverOutline();
            const hoverOutline = document.querySelector(
                "[data-testid='visual-builder__hover-outline']"
            );
            expect(hoverOutline).toHaveAttribute('style');

            const customCursor = document.querySelector(
                `[data-testid="visual-builder__cursor"]`
            );

            expect(customCursor).toHaveAttribute('data-icon', 'number');
            expect(customCursor?.classList.contains("visible")).toBeTruthy();
        });
    });

    describe("number field (multiple)", () => {
        let container: HTMLDivElement;
        let firstNumberField: HTMLParagraphElement;
        let secondNumberField: HTMLParagraphElement;
        let visualBuilder: VisualBuilder;

        beforeEach(() => {
            container = document.createElement("div");
            container.setAttribute(
                "data-cslp",
                "all_fields.bltapikey.en-us.number_multiple_"
            );
            container.getBoundingClientRect = vi
                .fn()
                .mockReturnValue(mockDomRect.singleHorizontal());

            firstNumberField = document.createElement("p");
            firstNumberField.setAttribute(
                "data-cslp",
                "all_fields.bltapikey.en-us.number_multiple_.0"
            );
            firstNumberField.getBoundingClientRect = vi
                .fn()
                .mockReturnValue(mockDomRect.singleLeft());

            secondNumberField = document.createElement("p");
            secondNumberField.setAttribute(
                "data-cslp",
                "all_fields.bltapikey.en-us.number_multiple_.1"
            );
            secondNumberField.getBoundingClientRect = vi
                .fn()
                .mockReturnValue(mockDomRect.singleRight());

            container.appendChild(firstNumberField);
            container.appendChild(secondNumberField);
            document.body.appendChild(container);

            visualBuilder = new VisualBuilder();
        });

        afterEach(() => {
            visualBuilder.destroy();
        });

        test("should have outline and custom cursor", async () => {
            container.dispatchEvent(mousemoveEvent);
            await waitForHoverOutline();
            const hoverOutline = document.querySelector(
                "[data-testid='visual-builder__hover-outline']"
            );
            expect(hoverOutline).toHaveAttribute('style');

            const customCursor = document.querySelector(
                `[data-testid="visual-builder__cursor"]`
            );

            expect(customCursor).toHaveAttribute('data-icon', 'number');
            expect(customCursor?.classList.contains("visible")).toBeTruthy();
        });

        test("should have outline and custom cursor on individual instances", async () => {
            firstNumberField.dispatchEvent(mousemoveEvent);
            await waitForHoverOutline();
            const hoverOutline = document.querySelector(
                "[data-testid='visual-builder__hover-outline']"
            );
            expect(hoverOutline).toHaveAttribute('style');

            const customCursor = document.querySelector(
                `[data-testid="visual-builder__cursor"]`
            );

            expect(customCursor).toHaveAttribute('data-icon', 'number');
            expect(customCursor?.classList.contains("visible")).toBeTruthy();
        });
    });
});
