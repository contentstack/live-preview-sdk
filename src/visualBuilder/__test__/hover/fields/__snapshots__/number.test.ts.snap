// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`When an element is hovered in visual builder mode > number field (multiple) > should have outline and custom cursor 1`] = `
<div
  data-cslp="all_fields.bltapikey.en-us.number_multiple_"
>
  <p
    data-cslp="all_fields.bltapikey.en-us.number_multiple_.0"
  />
  <p
    data-cslp="all_fields.bltapikey.en-us.number_multiple_.1"
  />
</div>
`;

exports[`When an element is hovered in visual builder mode > number field (multiple) > should have outline and custom cursor 2`] = `
<div
  class="visual-builder__hover-outline visual-builder__hover-outline--unclickable go217287799 go1719569523"
  data-testid="visual-builder__hover-outline"
  style="top: 34px; left: 34px; width: 828px; height: 54.3984375px;"
/>
`;

exports[`When an element is hovered in visual builder mode > number field (multiple) > should have outline and custom cursor 3`] = `
<div
  class="go1688149084 visual-builder__cursor visible"
  data-icon="number"
  data-testid="visual-builder__cursor"
>
  <div
    class="visual-builder__cursor-wrapper"
  >
    <div
      class="visual-builder__cursor-pointer go3046085498"
    >
      <svg
        fill="none"
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0.0167948 0.602214C-0.0784404 0.246792 0.246794 -0.0784402 0.602216 0.0167948L15.6457 4.04767C16.0914 4.1671 16.1259 4.78625 15.6962 4.95447L8.17161 7.90048C8.04753 7.94906 7.94936 8.04723 7.90078 8.17131L4.95449 15.6962C4.78627 16.1259 4.16712 16.0914 4.04769 15.6457L0.0167948 0.602214Z"
          fill="#5D50BE"
        />
      </svg>
    </div>
    <div
      class="visual-builder__cursor-icon go4261356330"
    >
      <svg
        fill="none"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M6.46172 8.08704C6.45117 8.02147 6.42928 7.9597 6.39813 7.90379C6.35727 7.8301 6.3012 7.7682 6.23536 7.72082C6.16973 7.67343 6.09347 7.63985 6.01084 7.62433C5.94777 7.61236 5.88203 7.61108 5.81629 7.62183C5.78173 7.6274 5.74822 7.63611 5.71607 7.64767L3.47852 8.39352C3.18381 8.49176 3.02453 8.81031 3.12277 9.10503C3.22101 9.39975 3.53956 9.55903 3.83428 9.46079L5.3439 8.95758L5.3439 15.302H4.59375C4.28309 15.302 4.03125 15.5538 4.03125 15.8645C4.03125 16.1752 4.28309 16.427 4.59375 16.427H7.33594C7.6466 16.427 7.89844 16.1752 7.89844 15.8645C7.89844 15.5538 7.6466 15.302 7.33594 15.302H6.4689L6.4689 8.19029C6.4697 8.15613 6.46737 8.12159 6.46172 8.08704Z"
          fill="white"
          fill-rule="evenodd"
        />
        <path
          clip-rule="evenodd"
          d="M9.51562 9.32004C9.51562 8.35585 10.2973 7.57422 11.2614 7.57422H12.8054C13.6809 7.57422 14.3906 8.28395 14.3906 9.15945V10.9828C14.3906 11.6816 13.9922 12.3191 13.3642 12.6254L12.2817 13.1533C12.2739 13.1571 12.266 13.1607 12.258 13.1641L11.3359 13.562C10.6003 13.8794 10.0831 14.537 9.93348 15.302H13.8281C14.1388 15.302 14.3906 15.5538 14.3906 15.8645C14.3906 16.1752 14.1388 16.427 13.8281 16.427H9.36208C9.03756 16.427 8.77241 16.1679 8.76493 15.8435C8.73182 14.4088 9.57261 13.0976 10.8902 12.5291L11.8003 12.1364L12.8711 11.6142C13.1125 11.4965 13.2656 11.2514 13.2656 10.9828V9.15945C13.2656 8.90527 13.0596 8.69922 12.8054 8.69922H11.2614C10.9186 8.69922 10.6406 8.97717 10.6406 9.32004C10.6406 9.6307 10.3888 9.88254 10.0781 9.88254C9.76746 9.88254 9.51562 9.6307 9.51562 9.32004Z"
          fill="white"
          fill-rule="evenodd"
        />
        <path
          clip-rule="evenodd"
          d="M16.5938 7.62677C16.2831 7.62677 16.0312 7.87861 16.0312 8.18927C16.0312 8.49993 16.2831 8.75177 16.5938 8.75177H19.5938C19.6973 8.75177 19.7812 8.83572 19.7812 8.93927V11.4583H16.5938C16.2831 11.4583 16.0312 11.7101 16.0312 12.0208C16.0312 12.3315 16.2831 12.5833 16.5938 12.5833H19.7812V15.1023C19.7812 15.2059 19.6973 15.2898 19.5938 15.2898H16.5938C16.2831 15.2898 16.0312 15.5417 16.0312 15.8523C16.0312 16.163 16.2831 16.4148 16.5938 16.4148H19.5938C20.3186 16.4148 20.9062 15.8272 20.9062 15.1023V8.93927C20.9062 8.2144 20.3186 7.62677 19.5938 7.62677H16.5938Z"
          fill="white"
          fill-rule="evenodd"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`When an element is hovered in visual builder mode > number field (multiple) > should have outline and custom cursor on individual instances 1`] = `
<p
  data-cslp="all_fields.bltapikey.en-us.number_multiple_.0"
/>
`;

exports[`When an element is hovered in visual builder mode > number field (multiple) > should have outline and custom cursor on individual instances 2`] = `
<div
  class="visual-builder__hover-outline visual-builder__hover-outline--unclickable go217287799 go1719569523"
  data-testid="visual-builder__hover-outline"
  style="top: 51px; left: 51px; width: 27.7734375px; height: 20.3984375px;"
/>
`;

exports[`When an element is hovered in visual builder mode > number field (multiple) > should have outline and custom cursor on individual instances 3`] = `
<div
  class="go1688149084 visual-builder__cursor visible"
  data-icon="number"
  data-testid="visual-builder__cursor"
>
  <div
    class="visual-builder__cursor-wrapper"
  >
    <div
      class="visual-builder__cursor-pointer go3046085498"
    >
      <svg
        fill="none"
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0.0167948 0.602214C-0.0784404 0.246792 0.246794 -0.0784402 0.602216 0.0167948L15.6457 4.04767C16.0914 4.1671 16.1259 4.78625 15.6962 4.95447L8.17161 7.90048C8.04753 7.94906 7.94936 8.04723 7.90078 8.17131L4.95449 15.6962C4.78627 16.1259 4.16712 16.0914 4.04769 15.6457L0.0167948 0.602214Z"
          fill="#5D50BE"
        />
      </svg>
    </div>
    <div
      class="visual-builder__cursor-icon go4261356330"
    >
      <svg
        fill="none"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M6.46172 8.08704C6.45117 8.02147 6.42928 7.9597 6.39813 7.90379C6.35727 7.8301 6.3012 7.7682 6.23536 7.72082C6.16973 7.67343 6.09347 7.63985 6.01084 7.62433C5.94777 7.61236 5.88203 7.61108 5.81629 7.62183C5.78173 7.6274 5.74822 7.63611 5.71607 7.64767L3.47852 8.39352C3.18381 8.49176 3.02453 8.81031 3.12277 9.10503C3.22101 9.39975 3.53956 9.55903 3.83428 9.46079L5.3439 8.95758L5.3439 15.302H4.59375C4.28309 15.302 4.03125 15.5538 4.03125 15.8645C4.03125 16.1752 4.28309 16.427 4.59375 16.427H7.33594C7.6466 16.427 7.89844 16.1752 7.89844 15.8645C7.89844 15.5538 7.6466 15.302 7.33594 15.302H6.4689L6.4689 8.19029C6.4697 8.15613 6.46737 8.12159 6.46172 8.08704Z"
          fill="white"
          fill-rule="evenodd"
        />
        <path
          clip-rule="evenodd"
          d="M9.51562 9.32004C9.51562 8.35585 10.2973 7.57422 11.2614 7.57422H12.8054C13.6809 7.57422 14.3906 8.28395 14.3906 9.15945V10.9828C14.3906 11.6816 13.9922 12.3191 13.3642 12.6254L12.2817 13.1533C12.2739 13.1571 12.266 13.1607 12.258 13.1641L11.3359 13.562C10.6003 13.8794 10.0831 14.537 9.93348 15.302H13.8281C14.1388 15.302 14.3906 15.5538 14.3906 15.8645C14.3906 16.1752 14.1388 16.427 13.8281 16.427H9.36208C9.03756 16.427 8.77241 16.1679 8.76493 15.8435C8.73182 14.4088 9.57261 13.0976 10.8902 12.5291L11.8003 12.1364L12.8711 11.6142C13.1125 11.4965 13.2656 11.2514 13.2656 10.9828V9.15945C13.2656 8.90527 13.0596 8.69922 12.8054 8.69922H11.2614C10.9186 8.69922 10.6406 8.97717 10.6406 9.32004C10.6406 9.6307 10.3888 9.88254 10.0781 9.88254C9.76746 9.88254 9.51562 9.6307 9.51562 9.32004Z"
          fill="white"
          fill-rule="evenodd"
        />
        <path
          clip-rule="evenodd"
          d="M16.5938 7.62677C16.2831 7.62677 16.0312 7.87861 16.0312 8.18927C16.0312 8.49993 16.2831 8.75177 16.5938 8.75177H19.5938C19.6973 8.75177 19.7812 8.83572 19.7812 8.93927V11.4583H16.5938C16.2831 11.4583 16.0312 11.7101 16.0312 12.0208C16.0312 12.3315 16.2831 12.5833 16.5938 12.5833H19.7812V15.1023C19.7812 15.2059 19.6973 15.2898 19.5938 15.2898H16.5938C16.2831 15.2898 16.0312 15.5417 16.0312 15.8523C16.0312 16.163 16.2831 16.4148 16.5938 16.4148H19.5938C20.3186 16.4148 20.9062 15.8272 20.9062 15.1023V8.93927C20.9062 8.2144 20.3186 7.62677 19.5938 7.62677H16.5938Z"
          fill="white"
          fill-rule="evenodd"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`When an element is hovered in visual builder mode > number field > should have outline and custom cursor 1`] = `
<p
  data-cslp="all_fields.bltapikey.en-us.number"
/>
`;

exports[`When an element is hovered in visual builder mode > number field > should have outline and custom cursor 2`] = `
<div
  class="visual-builder__hover-outline visual-builder__hover-outline--unclickable go217287799 go1719569523"
  data-testid="visual-builder__hover-outline"
  style="top: 51px; left: 51px; width: 27.7734375px; height: 20.3984375px;"
/>
`;

exports[`When an element is hovered in visual builder mode > number field > should have outline and custom cursor 3`] = `
<div
  class="go1688149084 visual-builder__cursor visible"
  data-icon="number"
  data-testid="visual-builder__cursor"
>
  <div
    class="visual-builder__cursor-wrapper"
  >
    <div
      class="visual-builder__cursor-pointer go3046085498"
    >
      <svg
        fill="none"
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0.0167948 0.602214C-0.0784404 0.246792 0.246794 -0.0784402 0.602216 0.0167948L15.6457 4.04767C16.0914 4.1671 16.1259 4.78625 15.6962 4.95447L8.17161 7.90048C8.04753 7.94906 7.94936 8.04723 7.90078 8.17131L4.95449 15.6962C4.78627 16.1259 4.16712 16.0914 4.04769 15.6457L0.0167948 0.602214Z"
          fill="#5D50BE"
        />
      </svg>
    </div>
    <div
      class="visual-builder__cursor-icon go4261356330"
    >
      <svg
        fill="none"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M6.46172 8.08704C6.45117 8.02147 6.42928 7.9597 6.39813 7.90379C6.35727 7.8301 6.3012 7.7682 6.23536 7.72082C6.16973 7.67343 6.09347 7.63985 6.01084 7.62433C5.94777 7.61236 5.88203 7.61108 5.81629 7.62183C5.78173 7.6274 5.74822 7.63611 5.71607 7.64767L3.47852 8.39352C3.18381 8.49176 3.02453 8.81031 3.12277 9.10503C3.22101 9.39975 3.53956 9.55903 3.83428 9.46079L5.3439 8.95758L5.3439 15.302H4.59375C4.28309 15.302 4.03125 15.5538 4.03125 15.8645C4.03125 16.1752 4.28309 16.427 4.59375 16.427H7.33594C7.6466 16.427 7.89844 16.1752 7.89844 15.8645C7.89844 15.5538 7.6466 15.302 7.33594 15.302H6.4689L6.4689 8.19029C6.4697 8.15613 6.46737 8.12159 6.46172 8.08704Z"
          fill="white"
          fill-rule="evenodd"
        />
        <path
          clip-rule="evenodd"
          d="M9.51562 9.32004C9.51562 8.35585 10.2973 7.57422 11.2614 7.57422H12.8054C13.6809 7.57422 14.3906 8.28395 14.3906 9.15945V10.9828C14.3906 11.6816 13.9922 12.3191 13.3642 12.6254L12.2817 13.1533C12.2739 13.1571 12.266 13.1607 12.258 13.1641L11.3359 13.562C10.6003 13.8794 10.0831 14.537 9.93348 15.302H13.8281C14.1388 15.302 14.3906 15.5538 14.3906 15.8645C14.3906 16.1752 14.1388 16.427 13.8281 16.427H9.36208C9.03756 16.427 8.77241 16.1679 8.76493 15.8435C8.73182 14.4088 9.57261 13.0976 10.8902 12.5291L11.8003 12.1364L12.8711 11.6142C13.1125 11.4965 13.2656 11.2514 13.2656 10.9828V9.15945C13.2656 8.90527 13.0596 8.69922 12.8054 8.69922H11.2614C10.9186 8.69922 10.6406 8.97717 10.6406 9.32004C10.6406 9.6307 10.3888 9.88254 10.0781 9.88254C9.76746 9.88254 9.51562 9.6307 9.51562 9.32004Z"
          fill="white"
          fill-rule="evenodd"
        />
        <path
          clip-rule="evenodd"
          d="M16.5938 7.62677C16.2831 7.62677 16.0312 7.87861 16.0312 8.18927C16.0312 8.49993 16.2831 8.75177 16.5938 8.75177H19.5938C19.6973 8.75177 19.7812 8.83572 19.7812 8.93927V11.4583H16.5938C16.2831 11.4583 16.0312 11.7101 16.0312 12.0208C16.0312 12.3315 16.2831 12.5833 16.5938 12.5833H19.7812V15.1023C19.7812 15.2059 19.6973 15.2898 19.5938 15.2898H16.5938C16.2831 15.2898 16.0312 15.5417 16.0312 15.8523C16.0312 16.163 16.2831 16.4148 16.5938 16.4148H19.5938C20.3186 16.4148 20.9062 15.8272 20.9062 15.1023V8.93927C20.9062 8.2144 20.3186 7.62677 19.5938 7.62677H16.5938Z"
          fill="white"
          fill-rule="evenodd"
        />
      </svg>
    </div>
  </div>
</div>
`;
